nexus/db-queries/src/db/queries/vpc_subnet.rs

```diff
@@ -88,11 +122,64 @@ impl SubnetError {
                     format!("IP address range '{}' conflicts with an existing subnet", ip).as_str()
                 )
             },
+            SubnetError::SameIdWithDifferentIpBlock(e) => e,
             SubnetError::External(e) => e,
         }
     }
 }
 
+/// Generate a subquery that either selects an exact record by ID, or generates
+/// an error, if the record does not have the same IP block as the input.
+///
+/// ```sql
+/// SELECT
+///     CAST(IF(ipv*_block = <ip>, <ip_as_str>, 'different-ip-block') AS INET)
+/// FROM
+///     vpc_subnet
+/// WHERE
+///     id = <id> AND time_deleted IS NULL
```

Comment:
Not sure I'm actually requesting a change here, but:

The `time_deleted IS NULL` check here has me curious -- if someone did actually soft-delete a `vpc_subnet`, and then try to re-run this insertion query **using the same UUID**, would we *want* it to pass this check?

Maybe this doesn't matter, if the attempt to `INSERT` this record again would always fail with a unique violation outside this CTE, but figured I'd ask, because I'm not sure how this particular case would manifest

Reply:
Hmm, a good question. First, I agree it seems unlikely, since it would require the exact same ID as a deleted record. But I also think we'd see this query "work", but not actually insert the new record! That seems bad.

Here's the flow, I think:

- A caller inserts a VPC Subnet with the same ID as the soft-deleted one (the IP block doesn't matter, actually)
- This CTE returns NULL, because there isn't an **active** record that matches
- The other CTE returns the IP block of the new record to be inserted, because there's no **active** record in the VPC with an overlapping block.
- We would attempt to insert the record, which would raise the PK violation
- We _catch_ that and ignore it. But the row in the DB doesn't match the one we wanted to insert! For sure, the `time_deleted` value doesn't match, and it's also possible if that the other columns don't match either.

Reply:
So there's one option, based on something Dave proposed in a chat. We could actually back out these changes, and use the original implementation, with one catch. If we do fail the query in the original way, because there is an overlapping subnet already in the database, we could try to fetch a record with the PK of the provided argument. If that returns nothing, then the overlap was with a different subnet, and we fail the test as previously. If we do find such a record, we can compare it and either return a conflict or internal error if the data doesn't match, or return it if it does match.

It feels to me like this opens a race window, but I could be wrong about that.

Reply:
Ok, after much toil in the SQL shell, I think I have a way to write the query that will handle the following cases:

1. Active row, overlapping IP blocks, different PK -> **FAIL**
2. Active row, non-equal IP blocks, same PK -> **FAIL**
3. Active row, equal IP blocks, same PK -> **DO NOTHING**
4. Inactive row, no IP constraint, same PK -> **UPDATE**

Item (4) is what you brought up here: what happens if there's a previously-deleted row that happens to have the same PK? In the query in the last commit, we would attempt to insert the new data, catch the PK violation, and then return anyway. But the data in this case **may not match**, since we don't actually know what the existing record was, only that it had the same PK and was deleted.

I believe this catches all the cases we really want, including idempotent inserts (for the saga, the main point of this); failing if you happen to try to overwrite a live record with the same PK but different IP blocks; and catching the original intended behavior, inserting any record that overlaps with others in the same VPC.

Reply:
I've added the monstrosity in 462bfe634. I updated and expanded the tests a bit, and things seem to be working OK.